pkg/controller/replication_controller.go

```diff
@@ -131,181 +91,251 @@ func NewReplicationManager(kubeClient client.Interface) *ReplicationManager {
 		// Sort the pods in the order such that not-ready < ready, unscheduled
 		// < scheduled, and pending < running. This ensures that we delete pods
 		// in the earlier stages whenever possible.
-		sort.Sort(activePods(filteredList))
+		sort.Sort(activePods(filteredPods))
 
 		wait := sync.WaitGroup{}
 		wait.Add(diff)
 		for i := 0; i < diff; i++ {
 			go func(ix int) {
 				defer wait.Done()
-				rm.podControl.deletePod(controller.Namespace, filteredList[ix].Name)
+				if err := rm.podControl.deletePod(controller.Namespace, filteredPods[ix].Name); err != nil {
+					// Decrement the expected number of deletes because the informer won't observe this deletion
+					rm.expectations.DeletionObserved(&controller)
+				}
 			}(i)
 		}
 		wait.Wait()
 	}
-	if controller.Status.Replicas != numActivePods {
-		controller.Status.Replicas = numActivePods
-		_, err = rm.kubeClient.ReplicationControllers(controller.Namespace).Update(&controller)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
 }
 
-func (rm *ReplicationManager) synchronize() {
-	// TODO: remove this method completely and rely on the watch.
-	// Add resource version tracking to watch to make this work.
-	var controllers []api.ReplicationController
-	list, err := rm.kubeClient.ReplicationControllers(api.NamespaceAll).List(labels.Everything())
+// syncReplicationController will sync the rc with the given key if it's fulfilled its expectations.
+// Update it's Status even if it hasnt't. This function is not meant to be invoked concurrently with the same key.
+func (rm *ReplicationManager) syncReplicationController(key string) error {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing controller %q (%v)", key, time.Now().Sub(startTime))
+	}()
+
+	obj, exists, err := rm.controllerStore.Store.GetByKey(key)
+	if !exists {
+		glog.Infof("Replication Controller has been deleted %v", key)
+		return nil
+	}
 	if err != nil {
-		util.HandleError(fmt.Errorf("synchronization error: %v", err))
-		return
+		glog.Infof("Unable to retrieve rc %v from store: %v", key, err)
+		rm.queue.Add(key)
+		return err
 	}
-	controllers = list.Items
-	wg := sync.WaitGroup{}
-	wg.Add(len(controllers))
-	for ix := range controllers {
-		go func(ix int) {
-			defer wg.Done()
-			glog.V(4).Infof("periodic sync of %v", controllers[ix].Name)
-			err := rm.syncHandler(controllers[ix])
-			if err != nil {
-				util.HandleError(fmt.Errorf("error synchronizing: %v", err))
-			}
-		}(ix)
+	controller := *obj.(*api.ReplicationController)
+
+	podList, err := rm.podStore.Pods(controller.Namespace).List(labels.Set(controller.Spec.Selector).AsSelector())
+	if err != nil {
+		glog.Errorf("Error getting pods for rc %q: %v", key, err)
+		rm.queue.Add(key)
+		return err
+	}
+	// TODO: Do this in a single pass, or use an index.
+	filteredPods := filterActivePods(podList.Items)
+
+	if rm.expectations.CheckExpectations(&controller) {
```

Comment:
Currently we just won't act on that delete till we've seen as many adds as expected (or the expectations expire on ttl). It might be worth hand locking for that degree of responsiveness, but it should only really matter on high latency networks (otherwise the watch should mostly work out?), and if the network is laggy do we really care that much about responsiveness in those cases? I will give it some thought, I had initially wanted to avoid hand locking but it might be worth it. As you suggested I'll file a bug and send a follow up pr.